### DISCUSSION TITLE
NestJS with Angular & NX - How to do TypeORM migrations?

### DISCUSSION THREAD

**Discussion by Chowarmaan** [nrwl/nx#11007] [General]
I have a project working outside of NX, that uses NestJS to provide the API for the Angular application.  I am looking to move this into NX for all the benefits.  The tutorial for Angular, even includes working with NestJS.  However, the main part missing, that I cannot find any information on, is getting the TypeORM support working in the NX structure.

I can create my entities in my NestJS project and configure the system to use the database. This all works. However, I do not know how to get my database changes dumped as data migrations, using the TypeORM of NestJS, as the package.json is now in the root directory, and if there are multiple NestJS API projects, the basic typeorm command line to generate migrations does not know which project to use.

In NPM, I would have

    "scripts": {
        ...
        "typeorm": "node --require ts-node/register ./node_modules/typeorm/cli.js"
    }

Then from the command line, you would run: `npm run typeorm migration:generate -- -n migrationNameHere`

I believe the command needs to be moved into the project.json of the API projects, but moving the command line into there produces errors.  Adding a step for generating migrations:

	"generate-migration": {
			"builder": "@nrwl/workspace:run-commands",
			"outputs": [],
			"options": {
				"command": "ts-node --project tsconfig.app.json ../../node_modules/.bin/typeorm migration:generate ",
				"cwd": "apps/timesheet-api"
			}
		},

This would need someway to get some parameters  (the name of migration to generate), but even trying to simply call this with a `--help` in the project entry, command line above, returns an error:

    ../../node_modules/.bin/ts-node --project tsconfig.app.json ../../node_modules/.bin/typeorm migration:generate --help
    node_modules\.bin\typeorm:2
    basedir=$(dirname "$(echo "$0" | sed -e 's,\\,/,g')")
                  ^^^^^^^

     SyntaxError: missing ) after argument list
        at Object.compileFunction (node:vm:352:18)
        at wrapSafe (node:internal/modules/cjs/loader:1033:15)
        at Module._compile (node:internal/modules/cjs/loader:1069:27)
        at Module._extensions..js (node:internal/modules/cjs/loader:1159:10)
        at Object.require.extensions.<computed> [as .js] (\node_modules\ts-node\src\index.ts:1587:43)
        at Module.load (node:internal/modules/cjs/loader:981:32)
        at Function.Module._load (node:internal/modules/cjs/loader:827:12)
        at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:77:12)
        at phase4 (\node_modules\ts-node\src\bin.ts:579:12)
        at bootstrap (\node_modules\ts-node\src\bin.ts:85:10)

How do you go about adding in the ability to use migrations to export the database changes that are made, using the TypeORM and NestJS?

**Comment by AndersBKnudsen** [comment]
Any help on this one?

**Reply by Chowarmaan** [reply]
I moved to Mikro-ORM, migrations and seeders all work.

**Reply by AndersBKnudsen** [reply]
Thanks for your reply! 

Cool, I will look into it! 
Do you have a working example or guide you followed?

**Comment by Chowarmaan** [comment]
Mostly their site.  However, I did get my migrations and seeds working by changing into the main directory of my application within the Monorepo (`cd apps/my_app`) and then run the commands in there.

## Database Migration scripts

Initial migration requires empty schema and no migrations data.
	
	npx mikro-orm migration:create --initial

In the individual apps directory (IE: apps/timesheet-api) run:
	
	npx mikro-orm migration:create

Apply Migrations:

	npx mikro-orm migration:up

Roll Back:

	npx mikro-orm migration:down

Generate Seeder (Create a new seeder class):

  	npx mikro-orm seeder:create [seeder]

Apply Seeders (pre-load data)

	npx mikro-orm seeder:run -c [seeder]

**Comment by henrysachs** [comment]
has anyone achieved this? :D

**Reply by Chowarmaan** [reply]
My example for the mikro-orm works, but I have not tried to work any further on TypeORM.

**Comment by adamwdennis** [comment]
I have a nestjs app, and many libs.  I find this setup to be nice.

## apps/api/project.json

```json
// these targets live in my apps/api/project.json:

    "typeorm": {
      "executor": "nx:run-commands",
      "outputs": [],
      "options": {
        "commands": [
          {
            "command": "ts-node -r tsconfig-paths/register -O '{\"module\": \"commonjs\"}' --project apps/api/tsconfig.app.json node_modules/typeorm/cli"
          }
        ]
      }
    },
    "typeorm-migration-generate": {
      "executor": "nx:run-commands",
      "outputs": [],
      "options": {
        "commands": [
          {
            "command": "nx run api:typeorm migration:generate libs/your/path/to/migrations/{args.name} -d libs/your/path/to/typeorm.datasource.ts --pretty"
          }
        ]
      }
    },
    "typeorm-migration-create": {
      "executor": "nx:run-commands",
      "outputs": [],
      "options": {
        "commands": [
          {
            "command": "nx run api:typeorm migration:create libs/your/path/to/migrations/{args.name}"
          }
        ]
      }
    },
    "typeorm-migration-run": {
      "executor": "nx:run-commands",
      "outputs": [],
      "options": {
        "commands": [
          {
            "command": "nx run api:typeorm migration:run -d libs/your/path/to/typeorm.datasource.ts"
          }
        ]
      }
    }
 ```
 
 
## package.json Scripts
I have the following scripts in my package.json:

```json
    "typeorm:migration:run": "NODE_ENV='development' nx run api:typeorm-migration-run",
    "typeorm:migration:generate": "NODE_ENV='development' nx run api:typeorm-migration-generate",
    "typeorm:migration:create": "NODE_ENV='development' nx run api:typeorm-migration-create",
```

## Generating a Migration

This assumes you've made changes to your entities...

```bash
yarn typeorm:migration:generate --name HereIsAName
```

## Running Migrations

```bash
yarn typeorm:migration:run
```

**Reply by linhnguyeneverfit** [reply]
Great! I replaced `{args.name}` by `$npm_config_name`& finally it works for me

**Reply by ugurkellecioglu** [reply]
thanks, in my case, I needed to change npm script into 

`"typeorm:migration:create": "NODE_ENV='development' nx run api:typeorm-migration-create --name=${npm_config_name}"`
